File: src/learning_gem5/part2/my_object.cc
#include "learning_gem5/part2/my_object.hh"

#include "base/trace.hh"
#include "debug/MyDbg.hh"

namespace gem5 {

MyObject::MyObject(const MyObjectParams &params) :
    SimObject(params),
    event([this]{processEvent();}, name()),
    my_exit(params.my_exit_object),
    myName(params.name),
    latency(params.time_to_wait),
    timesLeft(params.number_of_fires)
{
    DPRINTF(MyDbg, "From a MyObject: MyDbg\n");
    panic_if(!my_exit, "must have a non-null MyExitObject");
}

void MyObject::processEvent(void)
{
    timesLeft--;
    DPRINTF(MyDbg, "MyObject: processing the event\n");

    if (timesLeft <= 0) {
        DPRINTF(MyDbg, "Done firing\n");
        my_exit->myExit(myName);
    } else {
        schedule(event, curTick() + latency);
    }
}

void MyObject::startup()
{
    schedule(event, latency);
}

}

File: src/learning_gem5/part2/MyObject.py
from m5.params import *
from m5.SimObject import SimObject

class MyObject(SimObject):
    type = 'MyObject'
    cxx_header = 'learning_gem5/part2/my_object.hh'
    cxx_class = 'gem5::MyObject'

    time_to_wait = Param.Latency("Time before firing the event")
    number_of_fires = Param.Int(1, "Number of times to fire")
    my_exit_object = Param.MyExitObject('my_exit_obj in my_object')

class MyExitObject(SimObject):
    type = 'MyExitObject'
    cxx_header = 'learning_gem5/part2/my_exit_object.hh'
    cxx_class = 'gem5::MyExitObject'

    buffer_size = Param.MemorySize('1kB',
            'Size of buffer to fill')
    write_bandwidth = Param.MemoryBandwidth('100MB/s',
            'Bandwidth to fill the buffer')
